### ISSUE TITLE
Create v1.0 tag

### ISSUE THREAD

**Issue by bogosj** [bogosj/tesla#16]
I'd like to consider a v1.0 tag for the library after #8 is closed, preferable after #15 is merged.

**Comment by bogosj** [comment]
We should document how to acquire the token file first.

**Comment by michaelharo** [comment]
If I have time I'd like to improve the tests and possibly change some data types in the structs before v1.0.

**Comment by bogosj** [comment]
Fine with me.

**Comment by andig** [comment]
Could we expose the login stuff as api? My use is integrated with another application, running a separate cmd is not an option. I've basically done the same as here- copied and integrated- but I feel that it really deserves its own auth api?

**Comment by bogosj** [comment]
@andig going to track that here: https://github.com/bogosj/tesla/issues/22. I don't want to gate tagging v1.0 of a usable API wrapper on that. Hiding both of our comments here.

I've just finished up linting cmd/..., would you be interested in refactoring out the parts you care about to /login and having /cmd/login depend on it?

**Comment by andig** [comment]
Should the error return for `NewClient` be removed before 1.0?

**Comment by bogosj** [comment]
re: oauth2.Config - going to leave that to @michaelharo to think about as he made those changes.

re: go.mod 1.16 - is the suggestion we drop that directive completely? I'll swap out the usage of ioutil in another PR.

**Comment by bogosj** [comment]
@andig I'm using ioutil.ReadFile which is replaced by os.ReadFile doesn't exist <1.16. I'm not clear on the convention of how fast it's expected people would update to 1.16. I'm OK declaring that this library only works in 1.16+ but I'll wait to hear from you and @michaelharo and @uhthomas before I swap out the usages of ioutil.

**Comment by andig** [comment]
As long as 1.16 is targeted we can swap the functions. I don‘t think theres any need to require it, could probably be 1.13. I wouldn‘t mind 1.16 as I love the new features even if we don‘t need them here.

**Comment by uhthomas** [comment]
I think 1.16 is reasonable. The Go compatibility promise makes this quite safe.

**Comment by bogosj** [comment]
https://github.com/bogosj/tesla/pull/24 as expected, testing <1.16 is broken. I'll remove them from the matrix.

**Comment by bogosj** [comment]
Discussion offline with @michaelharo:

* I'm going to rollback and continue supporting Go 1.13+. He's currently on 1.15 and the changes that are 1.16 only will break him. We can investigate removing ioutil in the future.
* As for the (*Client, error) return type, we're going to keep that for future compatibility in the case that we offer optional args that could cause the client creation to fail.

**Comment by michaelharo** [comment]
Should `BaseURL` and `StreamingURL` be exported?  I'm wondering if they should be part of the 1.0 API.

**Comment by uhthomas** [comment]
Maybe not exported as a field, but definitely should be settable via functional options. It's essential for black box unit testing, and integration tests.

**Comment by bogosj** [comment]
@michaelharo @uhthomas something like https://github.com/bogosj/tesla/pull/33?

**Comment by michaelharo** [comment]
Yeah, that's what I was thinking.

**Comment by andig** [comment]
I will send 2 more small PRs. One rename, one handling token expiry

**Comment by michaelharo** [comment]
I just noticed that my change to make struct member names more meaningful is potentially inconsistent.  Not sure if we should care...

It uses both `DriverFrontDoor` and `FrontDriverWindow`.  This was based on the order of the letters in the json struct, but perhaps it `Front`/`Back` or `Driver`/`Passenger` should always be ordered the same in the names without regard to if it's a `Door` or `Window`?

**Comment by bogosj** [comment]
I'm ambivalent to the inconsistency - the inconsistency matches the structure presented by the API. I'm going to tag HEAD as v1.0.
